### ISSUE TITLE
[Errno 24] Too many open files

### ISSUE THREAD

**Issue by VelizarVESSELINOV** [GoogleCloudPlatform/gsutil#359]
gsutil -o GSUtil:parallel_composite_upload_threshold=150M mp cp -r "more than 4444 files" gs://...

The challenge is that I got several times this message during the upload and execution is "non responsive". The only solution I can see to delete what I uploaded and try again with less options.

**Comment by thobrla** [comment]
What operating system are you running on?  It's possible that raising the maximum number of open files may help.
On Linux: `ulimit  -n 10000`
On OSX: `ulimit -n 10000 && launchctl limit maxfiles 10000`

Alternatively, you can reduce parallel_process_count and/or parallel_thread_count in your .boto configuration file to reduce the number of files being transferred in parallel.

**Comment by VelizarVESSELINOV** [comment]
Thanks, I'm using OS X.

Before change:
```
> launchctl limit
...
maxfiles    256            unlimited
```

Error without sudo:
```
> ulimit -n 10000 && launchctl limit mayflies 10000
Could not set resource limits: 1: Operation not permitted
```

After sudo command:
```
> sudo launchctl limit maxfiles 10000 unlimited
> launchctl limit
maxfiles    10000          10240
```
Is it normal that unlimited is now 10240?

**Comment by thobrla** [comment]
I think the exact behavior is specific to the OS X kernel.  Did this help with the "Too many open files" error?

**Comment by VelizarVESSELINOV** [comment]
With the maxfile limit increase, I have less errors ("Too many open files"), but still a few errors during the transfer, so I need to restart several times to be sure that all files are transferred.

Without -o GSUtil:parallel_composite_upload_threshold=150M is even more stable the transfer (except few broken pipe errors), but probably slower.

**Comment by thobrla** [comment]
You could also reduce the parallel_process_count and parallel_thread count as I described earlier to further mitigate "Too many open files."

Are the broken pipe errors fatal?  If so, please open a separate issue describing that problem.
